# Drill # 1

In [153]:
import pandas as pd
import numpy as np

# create dataframe with the kids' ages
names = ['Greg', 'Marcia', 'Peter', 'Jan', 'Bobby', 'Cindy', 'Cousin Oliver']
df = pd.DataFrame(index = names)
df['age'] = [14, 12, 11, 10, 8, 6, 8]

# I imported this from a previous drill, so that drill #4 won't break
def modes(data):
    frequency = {}
    common = []
    for c in data:
        if c in frequency:
            frequency[c] += 1
        else:
            frequency[c] = 1
    for f in frequency:
        if frequency[f] == max(frequency.values()) != min(frequency.values()):
            common.append(f)
    return sorted(common)

def stats_test():
    # Compute our statistics
    mean = round(sum(df['age']/len(df['age'])),2)
    median = np.median([df['age']])
    mode = modes(df['age'].tolist())
    variance = round(np.var(df.age),2)
    st_dev = round(np.std(df['age']),2)
    se = round(np.std(df['age'] ,ddof=1) / np.sqrt(len(df['age'])),2)
    
    # create dataframe with statistics
    summary = pd.DataFrame()
    summary['stats'] = ['mean', 'median', 'mode', 'variance', 'st_dev', 'se']
    summary['values'] = [mean, median, mode, variance, st_dev, se]
    
    return summary
  
stats_test()

stats values
0      mean   9.86
1    median     10
2      mode    [8]
3  variance   6.41
4    st_dev   2.53
5        se   1.03

# Drill # 2

1. Mean is the most useful estimate of central tendency in this dataset. The mode under-represents the higher values in our dataset and while the median isn't a bad representation, it does have a higher variance than the mean for this dataset.
2. The best estimate of variance for our dataset is variance, since it best represents how far away the upper and lower ends of the age spectrum are from the mean.

# Drill # 3

In [105]:
df.loc[['Cindy'], ['age']] = 7
stats_test()

stats  values
0      mean   10.00
1    median   10.00
2      mode    8.00
3  variance    5.43
4    st_dev    2.33
5        se    0.95

1. What changed - mean, variance, standard deviation, and standard error
2. What did not change - median, mode

# Drill # 4

In [156]:
# print(df)
# df.loc[['Cousin Oliver']] = [['Jessica']]
df.loc[['Cousin Oliver'],['age']] = 1
#print(df) # doesn't update the name yet
stats_test()

stats values
0      mean   8.86
1    median     10
2      mode     []
3  variance  16.12
4    st_dev   4.02
5        se   1.64

1. Mode is now meangingless, and because 1 is a large outlier, the median is now a better unit of central tendency.
2. Variance is much higher since the difference between 1 is much further from the mean that any of the other values, and it best captures the difference in values from the mean.

# Drill # 5

In [126]:
magazines = ['TV Guide', 'Entertainment Weekly', 'Pop Culture Today', 'SciPhi Phanatic']
df = pd.DataFrame(index = magazines)
df['fans'] = [.2,.23,.17,.05]
mean = round(sum(df['fans']/len(df['fans'])),2)
print(mean)

0.16


Since the population is all adult Americans, and we don't have any numbers of how many people read each magazine, or if they read multiple magazines, I'm assuming they have an equal population/ weight and are distinct categories. Based on these conditions, 16% of adult Americans are likely to watch the Brady Bunch.

Conversely, we could throw out SciPhi Phanatic since it's an outlier and "probably" a smaller group of people, and get an average of 20%. However, this figure over-represents the population since it disregards adult Americans (the population) that are uninterested in those topics, or do not subscribe to magazines. People that subscribe to these magazines are more likely than the average young American to watch the Brady Bunch.